File: android/app/src/main/java/com/memorypets/android/presentation/setup/SetupScreen.kt
package com.memorypets.android.presentation.setup

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Pets
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.memorypets.sync.local.VaultBlobStore
import com.memorypets.sync.prefs.AppPrefs
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun SetupScreen(
    onSetupDone: () -> Unit,
    viewModel: SetupViewModel = hiltViewModel()
) {
    val state by viewModel.state.collectAsState()
    Scaffold(topBar = { TopAppBar(title = { Text("欢迎使用 MemoryPets") }) }) { padding ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(padding)
                .padding(horizontal = 24.dp)
                .verticalScroll(rememberScrollState()),
            verticalArrangement = Arrangement.spacedBy(16.dp),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            Spacer(Modifier.height(24.dp))
            Icon(Icons.Default.Pets, contentDescription = null,
                modifier = Modifier.size(72.dp), tint = MaterialTheme.colorScheme.primary)
            Text("第一步：设置解锁密码", style = MaterialTheme.typography.titleLarge)

            var pw by rememberSaveable { mutableStateOf("") }
            var pw2 by rememberSaveable { mutableStateOf("") }
            var words by rememberSaveable { mutableStateOf("记一下, 保存, 查一下, 导出") }
            OutlinedTextField(
                value = pw, onValueChange = { pw = it },
                label = { Text("解锁密码（≥ 6 位）") },
                visualTransformation = PasswordVisualTransformation(),
                isError = pw.length in 1..5,
                modifier = Modifier.fillMaxWidth()
            )
            OutlinedTextField(
                value = pw2, onValueChange = { pw2 = it },
                label = { Text("再输一遍") },
                visualTransformation = PasswordVisualTransformation(),
                isError = pw2.isNotEmpty() && pw2 != pw,
                modifier = Modifier.fillMaxWidth()
            )
            OutlinedTextField(
                value = words, onValueChange = { words = it },
                label = { Text("初始暗语（逗号分隔，可改）") },
                modifier = Modifier.fillMaxWidth()
            )

            Text(
                "可选 — 连接云同步 Relay（浏览器 ↔ 手机实时同步）",
                style = MaterialTheme.typography.titleMedium
            )
            var serverUrl by rememberSaveable { mutableStateOf(state.serverUrl) }
            var username by rememberSaveable { mutableStateOf("") }
            var cloudPw by rememberSaveable { mutableStateOf("") }
            OutlinedTextField(
                value = serverUrl, onValueChange = { serverUrl = it },
                label = { Text("服务器地址（例如 https://sync.example.com）") },
                modifier = Modifier.fillMaxWidth()
            )
            OutlinedTextField(
                value = username, onValueChange = { username = it },
                label = { Text("云账号用户名") },
                modifier = Modifier.fillMaxWidth()
            )
            OutlinedTextField(
                value = cloudPw, onValueChange = { cloudPw = it },
                label = { Text("云账号密码（与解锁密码互相独立）") },
                visualTransformation = PasswordVisualTransformation(),
                modifier = Modifier.fillMaxWidth()
            )

            if (state.error.isNotBlank()) {
                Text(state.error, color = MaterialTheme.colorScheme.error)
            }

            Button(
                onClick = {
                    viewModel.submit(
                        masterPassword = pw.takeIf { it == pw2 && it.length >= 6 }?.toCharArray(),
                        codewords = words.split(',', '，', ' ').map { it.trim() }.filter(String::isNotBlank),
                        serverUrl = serverUrl,
                        username = username.ifBlank { null },
                        cloudPassword = cloudPw.ifBlank { null }?.toCharArray(),
                        onDone = onSetupDone
                    )
                },
                enabled = pw.length >= 6 && pw == pw2 && state.loading.not(),
                modifier = Modifier.fillMaxWidth().padding(top = 8.dp)
            ) {
                if (state.loading) CircularProgressIndicator(Modifier.size(20.dp))
                else Text("完成 Setup")
            }
            Spacer(Modifier.height(24.dp))
        }
    }
}

data class SetupUiState(
    val loading: Boolean = false,
    val error: String = "",
    val serverUrl: String = AppPrefsImpl_DEFAULT_SERVER_URL
)

private const val AppPrefsImpl_DEFAULT_SERVER_URL = "https://sync.example.com"

@HiltViewModel
class SetupViewModel @Inject constructor(
    private val prefs: AppPrefs,
    private val blobStore: VaultBlobStore
) : ViewModel() {
    private val _state = MutableStateFlow(SetupUiState())
    val state: StateFlow<SetupUiState> = _state

    fun submit(
        masterPassword: CharArray?,
        codewords: List<String>,
        serverUrl: String,
        username: String?,
        cloudPassword: CharArray?,
        onDone: () -> Unit
    ) = viewModelScope.launch {
        if (masterPassword == null) {
            _state.value = _state.value.copy(error = "解锁密码至少 6 位且两次一致")
            return@launch
        }
        _state.value = _state.value.copy(loading = true, error = "")
        runCatching {
            prefs.setServerUrl(serverUrl)
            prefs.setCodewords(codewords)
            if (!username.isNullOrBlank() && cloudPassword != null) {
                // TODO: 调用 CloudSyncApi.register 或 login；成功后 setCloudAccount(username, token)
            }
        }.onFailure { t ->
            _state.value = _state.value.copy(error = t.message ?: "setup failed")
        }.onSuccess { onDone() }
        _state.value = _state.value.copy(loading = false)
    }
}

File: flutter/lib/screens/setup_screen.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';

import '../api/cloud_sync_api.dart';
import '../session/vault_session.dart';
import '../storage/app_prefs.dart';

class SetupScreen extends StatefulWidget {
  final VoidCallback onSetupDone;
  const SetupScreen({super.key, required this.onSetupDone});

  @override
  State<SetupScreen> createState() => _SetupScreenState();
}

class _SetupScreenState extends State<SetupScreen> {
  final _pwCtrl = TextEditingController();
  final _pw2Ctrl = TextEditingController();
  final _wordsCtrl = TextEditingController(text: '记一下, 保存, 查一下, 导出');
  final _serverCtrl = TextEditingController();
  final _userCtrl = TextEditingController();
  final _cloudPwCtrl = TextEditingController();

  bool _loading = false;
  String _error = '';

  @override
  void initState() {
    super.initState();
    final prefs = context.read<AppPrefs>();
    _serverCtrl.text = prefs.serverUrl;
  }

  @override
  void dispose() {
    _pwCtrl.dispose();
    _pw2Ctrl.dispose();
    _wordsCtrl.dispose();
    _serverCtrl.dispose();
    _userCtrl.dispose();
    _cloudPwCtrl.dispose();
    super.dispose();
  }

  bool get _canSubmit =>
      _pwCtrl.text.length >= 6 && _pwCtrl.text == _pw2Ctrl.text && !_loading;

  Future<void> _submit() async {
    if (!_canSubmit) {
      setState(() => _error = '解锁密码至少 6 位且两次一致');
      return;
    }
    setState(() {
      _loading = true;
      _error = '';
    });
    final prefs = context.read<AppPrefs>();
    final session = context.read<VaultSession>();
    try {
      await prefs.setServerUrl(_serverCtrl.text.trim());
      final words = _wordsCtrl.text
          .split(RegExp('[,，\\s]'))
          .map((e) => e.trim())
          .where((e) => e.isNotEmpty)
          .toList();
      await prefs.setCodewords(words);

      final username = _userCtrl.text.trim();
      final cloudPw = _cloudPwCtrl.text;
      if (username.isNotEmpty && cloudPw.isNotEmpty) {
        // 云同步是可选项：这里的任何失败（网络不可达 / 域名解析失败 / 账号错误）
        // 都只应提示用户，不能阻塞本地 Setup 完成。
        try {
          final api = CloudSyncApi(prefs);
          var resp = await api.login(username, cloudPw);
          if (!resp.ok) {
            resp = await api.register(username, cloudPw);
          }
          if (resp.ok && resp.token != null) {
            await prefs.setCloudAccount(username, resp.token!);
          } else {
            setState(() => _error = '云账号登录/注册失败：${resp.error ?? "未知错误"}（本地 Setup 仍会继续）');
          }
        } catch (t) {
          setState(() => _error = '云同步暂时无法连接：$t（本地 Setup 仍会继续，可稍后在设置里重试）');
        }
      }

      await session.createNew(_pwCtrl.text);
      widget.onSetupDone();
    } catch (t) {
      setState(() => _error = t.toString());
    } finally {
      if (mounted) setState(() => _loading = false);
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('欢迎使用 MemoryPets')),
      body: SingleChildScrollView(
        padding: const EdgeInsets.symmetric(horizontal: 24),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.center,
          children: [
            const SizedBox(height: 24),
            Icon(Icons.pets, size: 72, color: Theme.of(context).colorScheme.primary),
            const SizedBox(height: 16),
            Text('第一步：设置解锁密码', style: Theme.of(context).textTheme.titleLarge),
            const SizedBox(height: 16),
            TextField(
              controller: _pwCtrl,
              obscureText: true,
              decoration: const InputDecoration(labelText: '解锁密码（≥ 6 位）'),
              onChanged: (_) => setState(() {}),
            ),
            const SizedBox(height: 12),
            TextField(
              controller: _pw2Ctrl,
              obscureText: true,
              decoration: const InputDecoration(labelText: '再输一遍'),
              onChanged: (_) => setState(() {}),
            ),
            const SizedBox(height: 12),
            TextField(
              controller: _wordsCtrl,
              decoration: const InputDecoration(labelText: '初始暗语（逗号分隔，可改）'),
            ),
            const SizedBox(height: 24),
            Text('可选 — 连接云同步 Relay（浏览器 ↔ 手机实时同步）',
                style: Theme.of(context).textTheme.titleMedium),
            const SizedBox(height: 12),
            TextField(
              controller: _serverCtrl,
              decoration: const InputDecoration(labelText: '服务器地址（例如 https://sync.example.com）'),
            ),
            const SizedBox(height: 12),
            TextField(
              controller: _userCtrl,
              decoration: const InputDecoration(labelText: '云账号用户名'),
            ),
            const SizedBox(height: 12),
            TextField(
              controller: _cloudPwCtrl,
              obscureText: true,
              decoration: const InputDecoration(labelText: '云账号密码（与解锁密码互相独立）'),
            ),
            if (_error.isNotEmpty) ...[
              const SizedBox(height: 12),
              Text(_error, style: TextStyle(color: Theme.of(context).colorScheme.error)),
            ],
            const SizedBox(height: 16),
            SizedBox(
              width: double.infinity,
              child: FilledButton(
                onPressed: _canSubmit ? _submit : null,
                child: _loading
                    ? const SizedBox(
                        width: 20, height: 20, child: CircularProgressIndicator(strokeWidth: 2))
                    : const Text('完成 Setup'),
              ),
            ),
            const SizedBox(height: 24),
          ],
        ),
      ),
    );
  }
}
